File: Toocal/Meta.fs
module Toocal.Core.DataAccessLayer.Meta

open System
open Toocal.Core.DataAccessLayer.Page
open Toocal.Core.ISerializable

type Meta() =

  let mutable _FreelistPage: PageNum = 0UL

  member this.FreelistPage
    with get () = _FreelistPage
    and set (freelistPage: PageNum) = _FreelistPage <- freelistPage

  static member public META_PAGE_NUM: PageNum = 0UL

  interface ISerializable with
    member this.Serialize(buffer: Byte[]) =
      let serialized_freelist = BitConverter.GetBytes this.FreelistPage
      Array.blit serialized_freelist 0 buffer 0 serialized_freelist.Length

    member this.Deserialize(buffer: Byte[]) =
      this.FreelistPage <- BitConverter.ToUInt64 buffer

  member public this.Serialize = (this :> ISerializable).Serialize
  member public this.Deserialize = (this :> ISerializable).Deserialize

File: src/main/scala/com/muqiuhan/toocal/dal/Meta.scala
/* Copyright (c) 2023 Muqiu Han
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright notice,
 *       this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright notice,
 *       this list of conditions and the following disclaimer in the documentation
 *       and/or other materials provided with the distribution.
 *     * Neither the name of Terifo nor the names of its contributors
 *       may be used to endorse or promote products derived from this software
 *       without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.muqiuhan.toocal.dal

import java.nio.ByteBuffer

/** Meta is a special Page whose PageNum is always 0. It is the first page of the database and is used to store metadata
  * in the database, such as the location of FreeList and other information.
  */
case class Meta(
    /** the page of FreeList should be on the next page of Meta */
    var freeListPage: PageNum = 1
):

    def serialize(buffer: Array[Byte]): Unit = ByteBuffer.allocate(Page.PAGE_NUM_SIZE).putLong(freeListPage).array
        .copyToArray(buffer)

    def deserialize(buffer: Array[Byte]): Unit = freeListPage = ByteBuffer.wrap(buffer).getLong()

object Meta:
    val PAGE_NUM: Int = 0
